corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js

```diff
@@ -248,6 +250,11 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             } else if (this.model.get('input') === 'address') {
                 // geocoderItemCallback sets the value on the model
             } else {
+                if (!this.isValid()) {
```

Comment:
Question about this, because I'm starting to look at the conditionally required fields work, [USH-2203](https://dimagi-dev.atlassian.net/browse/USH-2203), which overlaps with this work pretty heavily:

In [USH-2225](https://dimagi-dev.atlassian.net/browse/USH-2225), you mention that you went with doing validation on input change, not submit, but I don't see quite how that's happening. You need to make a request to formplayer to validate the field, right? `isValid` just looks at the current state of the model. There's a formplayer request in `changeDropdown` (so named because currently its only job is to determine if changing a dropdown caused the options in a different dropdown to change), but that only runs if the field that changed is a dropdown.

I think you need to do something like [this](https://github.com/dimagi/commcare-hq/compare/jls/wip) to call `changeDropdown` when any field changes and then, inside of `changeDropdown`, check all fields for validity and update error messages.

If that sounds right, lmk if/how you'd like me to contribute, since this is bleeding into the conditionally required fields work and I think you have more going on this sprint than I do.

Reply:
Thanks Jenny, that clears up some things for me in terms of getting real-time validation working on input change. With that said, I'm now thinking I should backtrack on implementing validation on input change and stick with the original scope for this ticket, on submit validation. Considering that it's already starting to overlap with other planned validation work and the amount of other work I personally have this sprint I think that makes sense. Thoughts?

Reply:
Yeah, that sounds good. This work can validate on submit, and that's easy to describe for QA, and then validation on change can be part of the dynamic conditions work that I'll pick up this sprint.